packages/core/src/http-handler.ts

```diff
@@ -45,6 +51,7 @@ export class HttpHandler {
 
     this._bindFindRoute(requestContext);
     this._bindInvokeMethod(requestContext);
+    this._bindAuthenticationFn(requestContext);
```

Comment:
IMO, the authentication function binding should be provided by `@loopback/authentication` component and registered in the application context (container) at the time when the authentication component is added to the app. I believe that's why we added the concept of providers, right?

Note that bindings are inherited: you can bind a value to application and then retrieve it from per-request context. If the binding has any dependencies, then they will be injected using the values provided by the per-request context.

```ts
class AuthenticationFnProvider {
  constructor(
    @inject(BindingKeys.Context.CONTROLLER_METHOD_META) private meta: AuthenticationMetadata,
     // ...)
  
  value() {
    return async (request: ParsedRequest) => {
      // ...
    };
  }
}
```

As I am reading the implementation of your "authenticateFn" function, I see that you need to resolve context bindings and also define new bindings, all in the same function. IIRC, our current Context does not support such usage, so we will need to figure out an elegant way for enabling it.

The easiest solution (but also the worst one IMO) is to make the context injectable, i.e. add the following line to `Context` constructor:

```ts
  constructor(private _parent?: Context) {
    this.registry = new Map();
    this.bind('context').to(this);
  }
```

I would personally prefer to expose a more fine-grained API, e.g.

```ts
// inside Context constructor
  this.bind('bind').to(key => this.bind(key));
  this.bind('inject').to(key => this.get(key));

// usage
class MyProvider {
  constructor(
    @inject('inject') private inject: (key: string) => Promise<BoundValue>,
    @inject('bind') private bind: (key: string) => Binding) {}
  value() {
    const meta: AuthenticationMetadata = await inject(CONTROLLER_METHOD_META
    // ...
    bind(AUTHENTICATION_STRATEGY).to(strategy);
  }
}
```

@raymondfeng @ritch @superkhau thoughts?